include/deal.II/base/patterns.h

```diff
@@ -0,0 +1,1681 @@
+       *
+       * @author Luca Heltai, 2017
+       */
+      template <class T, class Enable = void>
+      struct RankInfo
+      {
+        static constexpr int list_rank = 0;
+        static constexpr int map_rank = 0;
+      };
+    }
+
+    // Arithmetic types
+    template <class T>
+    struct Convert<T, typename std::enable_if<std::is_arithmetic<T>::value>::type>
+    {
+
+      static std::unique_ptr<Patterns::PatternBase> to_pattern()
+      {
+        if (std::is_integral<T>::value)
+          return std_cxx14::make_unique<Patterns::Integer>(
+                   std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+        else if (std::is_floating_point<T>::value)
+          return std_cxx14::make_unique<Patterns::Double>(
+                   -std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
+      }
+
+      static std::string to_string(const T &value,
+                                   const std::unique_ptr<Patterns::PatternBase>
+                                   &p = Convert<T>::to_pattern())
+      {
+        std::string str;
+        if (std::is_same<T, unsigned char>() || std::is_same<T, char>())
+          str = std::to_string((int)value);
+        else
+          str = std::to_string(value);
+        AssertThrow(p->match(str), ExcNoMatch(str, *p));
+        return str;
+      }
+
+      static T to_value(const std::string &s,
+                        const std::unique_ptr<Patterns::PatternBase> &p =
+                          Convert<T>::to_pattern())
+      {
+        AssertThrow(p->match(s), ExcNoMatch(s, *p));
+        std::istringstream is(s);
+        T value;
+        if (std::is_same<T, unsigned char>::value || std::is_same<T, char>::value)
+          {
+            int i;
+            is >> i;
+            value = i;
+          }
+        else
+          is >> value;
+
+        // If someone passes "123 abc" to the function, the method yelds an
+        // integer 123 alright, but the space terminates the read from the string
+        // although there is more to come. This case, however, is checked for in
+        // the call p->match(s) at the beginning of this function, and would
+        // throw earlier. Here it is safe to assume that if we didn't fail the
+        // conversion with the operator >>, then we are good to go.
+        AssertThrow(!is.fail(),
+                    ExcMessage("Failed to convert from \"" + s +
+                               "\" to the type \"" +
+                               boost::core::demangle(typeid(T).name()) + "\""));
+        return value;
+      }
+    };
+
+    namespace internal
+    {
+      const std::array<std::string, 4> default_list_separator {{","  ,  ";"  ,  "|"  ,   "%"}};
+      const std::array<std::string, 4> default_map_separator {{":"  ,  "="  ,  "@"  ,   "#"}};
+
+      //specialize a type for all of the STL containers and maps
+      template <typename T>       struct is_stl_container : std::false_type {};
```

Comment:
I would have expected that a map is also an ``stl_container``. Is the naming a bit unfortunate?